### ISSUE TITLE
Getting _PARTITION_EOF error when using consume

### ISSUE THREAD

**Issue by bkim54** [ah-/rdkafka-dotnet#87]
I am using the Consumer.Consume method with a 30 sec timeout (because I want to receive just 1 message) however randomly (about 1/10 times) the returned MessageAndError contains the error "_PARTITION_EOF"

this is how I set up my consumer
```
Config kafkaConfig = new Config() 
 {
      GroupId = replyInboxString, // this is a random guid-string
       DefaultTopicConfig = new TopicConfig { ["auto.offset.reset"] = "smallest" },
       EnableAutoCommit = true
};

```
Then I use assign and set each offset for each partition for my one topic to be Offset.Beginning

My cluster is setup with 3 partitions.

am I doing something wrong? why do I get this error randomly?

**Comment by ah-** [comment]
It's not really an error, it's just a notification that your consumer has reached the end of that partition and there aren't any other messages immediately available on it. If you're not interested in it you can just ignore it.

**Comment by bkim54** [comment]
I am trying to create a request/reply wrapper with RdKafka, I am expecting one message to be consumed from the topic, I understand that only one partition will have the message, is this "_PARTITION_EOF" coming from the other 2 partitions then?

**Comment by treziac** [comment]
Nope, it will come from the partition from which you just consumed the last message. Or at start, from empty partition or with no new message

Example: let's assume you are consuming and there is no new message in any partition.
You are then trying to consume, and timeout every 30s, which means a null message is return from Consume (and typically just ignore it, look at eventConsumer for example)

You then push 3 messages, 2 to P0 and 1 to P1 (assuming 3 partitions)
Your loop will then:
- A1 : Consume and return immediately message 1 from P0
- B1 : Consume and return immediately message 1 from P1
- A2 : Consume and return immediately message 2 from P0
- A3 : Consume and return immediately EOF from P0 (as there is no more available message in the internal queue)
- B2 : Consume and return immediately EOF from P1
- Consume and wait for timeout (no more message available)

Note that A and B (A for P0 and B for P1) may be mixed in whatever order (A1/A2/A3/B1/B2 or B1/A1/A2/B2/B3 are other possible flow, there is one thread for each partition internally)

In an application where you receive lot of messages but can easily keep up with the throughput, you will receive lot of EOF. It just means you are not cowling under request.
In a topic with lot of partition which are partitioned, when you process lot of past data from kafka, it allow you to know when all data from a partition have been received so you can process them immediately.

In your use case, it seems useless :) If you are not interested by eof notifications, you can either ignore them or desactivate them via conf (conf["enable.partition.eof"] = "false")

By the way ( @ah- correct me if i'm wrong) you should stick with a 1s timeout as consume also triggers errors (transport error for example) so you may want to get them immediately and not wait for a timeout (i'm not entirely sure with this)

**Comment by bkim54** [comment]
ahh I see thanks for the explanation @treziac.

on another note, i mentioned I was using Assign to reset each topics partitions to Offset.Beginning, however, When use subscribe like so
`                consumer.Subscribe(new List<string> { replyInboxString });
`

The consumer does not get any messages sometimes. what is the difference between using Assign and Consume vs Subscribe and Consume?

**Comment by treziac** [comment]
oups, didn't post this when getting off, still post it for info :)

Subscribe use group.id, group coordinator and rebalance mechanism (see here for more explanations)
https://www.safaribooksonline.com/library/view/kafka-the-definitive/9781491936153/ch04.html

TL;DR : 
Subscribe put consumer in a group depending of it's group id. All consumers from a group share the data (if a topic has 3 partitions and there is two consumers from same group subscribing to this topic, one will receive message from 2 of th partition and the second from the third partition)
If you don't close consumer properly, group coordinator don't know the consumer exited the group until session timeout (30s by default for librdkafka)
Also that's why you shouldn't put a 30s timeout - you could just be ejected from group as you don't send hearthbate, so you may have to wait an other 30s to get a rebalance callback... 
If you run multiple consumer and you want all of them to receive message, you can give them a different group id (still not possible to use subscribe with no group id)


Assign just assign directly the partition to the consumer. No group coorinator. Also no discovery of new partition, sharing or other. But you will consume immediatly and are not impacted by other consumer

Note that you can use OnPartition Assigned/Revoked to see what is happening (check the advanced consumer example)
